Add package wtxmgr to use walletdb interface

tuxcanfly opened the pull request:
Started rolling out the database changes for migrating to boltdb, added db row for Block. Work in progress.

davecgh committed:
Add conversion from legacy txstore if it exists.

davecgh committed:
Move keystore to legacy directory.

tuxcanfly committed:
Turn echo off when reading passwords

davecgh committed:
Switch to new waddrmgr package

This commit converts the wallet to use the new secure hierarchical
deterministic wallet address manager package as well as the walletdb
package.

The following is an overview of modified functionality:

- The wallet must now be created before starting the executable
- A new flag --create has been added to create the new wallet using wizard
  style question and answer prompts
- Starting the process without an existing wallet will instruct now
  display a message to run it with --create
- Providing the --create flag with an existing wallet will simply show an
  error and return

tuxcanfly commented:
I have `wtxmgr` now working completely with `walletdb`'s boltdb database. Now wrapping up the legacy txstore import; will rebase, cleanup and put this up for review soon.

davecgh commented:
Great.  Can you put the pull request against the conformal repo?  Just mention that is depends on 147.

tuxcanfly commented:
Sure, will do that.

davecgh committed:
Return the memory after scrypt to the OS.

Previously a runtime.GC was being invoked which forced it to release the
memory as far as the garbage collector is concerned, but the memory was
not released back to the OS immediatley.  This modification allows the
memory to be released immedately since it won't be needed again until the
next wallet unlock.

jrick committed:
Implement several btcd RPCs in wallet as well.

Wallet should handle these so that wallet clients don't end up
creating requests to btcd about the latest processed block, which is
not the same as wallet's most recently processed block.

By providing wallet clients with this info, we avoid a race where the
client thinks wallet has processed N blocks, but in fact is still
synced to N-1 (and perhaps currently processing transactions from
block N).  This can cause unexpected results for many of the
bitcoind-compatible RPC APIs due to their reliance on number of
confirmations, rather than using absolute block heights.

jrick committed:
Use the correct synced block when calculating confs.

This was previously using the most recently notified (by the chain
package) block, but transaction processing from this block may not be
finished yet.  Using this block's height to calculate the number of
confirmations is therefore incorrect, and can result in every RPC
handler missing transactions or returning transactions from the wrong
block.

tuxcanfly committed:
Move txstore package to legacy directory

tuxcanfly committed:
Add pkg wtxmgr - new txstore with walletdb backend

tuxcanfly committed:
Updated wallet to use wtxmgr

jrick committed:
Prevent deadlock when changing passphrases.

davecgh committed:
Update btcwire path import paths to new location.

jrick committed:
Fix Go versions for Travis.

jrick committed:
Use Travis' container-based infrastructure.

jrick committed:
Address several issues pointed out by lint and vet.

This brings the entire tree closer to but not 100% goclean.sh clean.

davecgh committed:
Update btcnet path import paths to new location.

jrick committed:
Merge branch 'master' into intwaddrmgr

Conflicts:
	btcwallet.go
	chain/chain.go
	wallet.go

davecgh committed:
Update btcec path import paths to new location.

jrick committed:
Merge branch 'master' into intwaddrmgr

davecgh committed:
waddrmgr: Update documentation for walletdb.

This commit updates the documentation which discusses creating and opening
the manager to properly mention the wallet database namespace as well as
another typo.

davecgh committed:
waddrmgr: Update more documentation for walletdb.

tuxcanfly committed:
Move rename to legacy

tuxcanfly committed:
Address review items:

* remove unnecessary if err := fn(); err != nil branches
* use n+1, n-1 in arg instead of n++, n--
* handle index out of range due to invalid metadata
* remove unncessary shaHashToBytes
* format err message for deserializeOutPoint
* remove unreachable code in fetchNumBlockTxRecords
* Return nil in place of empty array on err

tuxcanfly committed:
Address review items:

* update wtxmgr doc.go for txstore to wtxmgr migration

jrick committed:
Merge branch 'master' into intwaddrmgr

tuxcanfly committed:
Use a debits bucket for storing tx debits

tuxcanfly committed:
Use debit bucket for both confirmed/unconfirmed tx

tuxcanfly committed:
Use credit bucket for both unconfirmed/confirmed tx

tuxcanfly committed:
Update only credits/debits bucket on AddCredit/AddDebit

davecgh committed:
Update btcjson path import paths to new location.

davecgh committed:
Update btcws path import paths to new location.

jrick committed:
Use rescan for initial tx notification registrations.

Requires btcsuite/btcd#310.

jrick committed:
Merge branch 'master' into intwaddrmgr

Conflicts:
	chain/chain.go
	rpcserver.go

tuxcanfly committed:
Use single bucket for all tx credits

jrick committed:
Fix build.

This fixes the build after btcsuite/btcd#302, which renamed the
txscript constant ScriptCanonicalSignatures constant to
ScriptVerifyDERSignatures.

jrick committed:
Merge branch 'master' into intwaddrmgr

jrick committed:
Merge branch 'master' into intwaddrmgr

Conflicts:
	wallet.go

tuxcanfly committed:
Use TxStoreError for all errors returned by wtxmgr

tuxcanfly committed:
Update single credit at a time

tuxcanfly committed:
Address review items:

* remove unnecessary valid pointer funcs
* simplify bool serialization

tuxcanfly committed:
Address review items:

* define serialize funcs before deserialize - outpoint, block
* pass in object ptrs for deserialize funcs - outpoint, block

tuxcanfly committed:
Address review items:

* order all serialize/deserialize funcs

tuxcanfly committed:
Use meta bucket for maintaining block tx count

tuxcanfly committed:
Address review items:

* style change - func declarations
* txrecord - return out ptr on deserialize

tuxcanfly committed:
Address review items:

* formatting, %s for err message

tuxcanfly committed:
Use fixed size buf for serializing credit

tuxcanfly committed:
Added block, txrecord serialization benchmarks

tuxcanfly committed:
Updated doc and style for db funcs

tuxcanfly committed:
Utilize buffer reuse in serialization
